tests/api/auth_test.py

```diff
@@ -0,0 +1,59 @@
+import pytest
+
+from zino.api.auth import AuthenticationFailure, authenticate, get_challenge, read_users
+
+
+class TestAuthenticate:
+    def test_when_challenge_response_is_ok_should_return_true(self, secrets_file):
+        good_response = "b3e9ba156949152c5c6f3334cf86333b1daabfe9"
+        assert authenticate(
+            "user1",
+            response=good_response,
+            challenge="a97fbebb3eef2dfcf167645229c0c5fa9e92e3da",
+            secrets_file=secrets_file,
+        )
+
+    def test_when_challenge_response_is_bad_should_raise_error(self, secrets_file):
+        bad_response = "aaa9ba156949152c5c6f3334cf86333b1daabbbb"
+        with pytest.raises(AuthenticationFailure):
+            assert authenticate(
+                "user1",
+                response=bad_response,
+                challenge="a97fbebb3eef2dfcf167645229c0c5fa9e92e3da",
+                secrets_file=secrets_file,
+            )
+
+    def test_when_user_does_not_exist_should_raise_error(self, secrets_file):
+        with pytest.raises(AuthenticationFailure):
+            assert authenticate(
+                "nobody",
+                response="doesntmatter",
+                challenge="doesntmatter",
+                secrets_file=secrets_file,
+            )
+
+    def test_when_challenge_is_empty_should_raise_error(self, secrets_file):
+        """Can be removed when challenge-less auth is supported"""
+        with pytest.raises(AuthenticationFailure):
+            assert authenticate(
+                "user1",
+                response="doesntmatter",
+                secrets_file=secrets_file,
+            )
+
+
+class TestGetChallenge:
+    def test_should_never_return_same_challenge(self):
+        count = 10
+        challenges = set(get_challenge() for _ in range(count))
+        assert len(challenges) == 10, "same challenge was produced more than once"
```

Comment:
This is a weird test imo.

While its unlikely to generate the same value, the `secrets` library does not actually promise to not generate the same value twice. Promising unique values would technically make tokens less secure (or more predictable), as this would mean one possible number is removed every time its run.

The `get_challenge` wrapping function also does nothing to avoid duplicates. So this test technically tests for something that is not actually expected behavior.

Also, this test is technically "flaky", since there is a rare possibility that you actually get duplicates

Reply:
This test is mostly meant to safeguard against a n00b developer b0rking the `get_challenge()` implementation 10 years down the line into something that returns the same static value every time it is called.  It is not meant to be a spec that says that challenges should *NEVER* be the same.

I will personally buy you a very expensive craft beer if you ever observe the secrets library to generate the same value twice in a row during this test :)

Reply:
> It is not meant to be a spec that says that challenges should _NEVER_ be the same.

then you should rename the test to not be `test_should_never_return_same_challenge`, Theres not really any other way to interpret "never return same challenge"

Reply:
done